src/graph/optimizer/rule/PushFilterDownTraverseRule.cpp

```diff
@@ -0,0 +1,164 @@
+/* Copyright (c) 2022 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License.
+ */
+
+#include "graph/optimizer/rule/PushFilterDownTraverseRule.h"
+
+#include "common/expression/ConstantExpression.h"
+#include "common/expression/Expression.h"
+#include "graph/optimizer/OptContext.h"
+#include "graph/optimizer/OptGroup.h"
+#include "graph/planner/plan/PlanNode.h"
+#include "graph/planner/plan/Query.h"
+#include "graph/util/ExpressionUtils.h"
+#include "graph/visitor/ExtractFilterExprVisitor.h"
+
+using nebula::Expression;
+using nebula::graph::Filter;
+using nebula::graph::PlanNode;
+using nebula::graph::QueryContext;
+using nebula::graph::Traverse;
+
+namespace nebula {
+namespace opt {
+
+std::unique_ptr<OptRule> PushFilterDownTraverseRule::kInstance =
+    std::unique_ptr<PushFilterDownTraverseRule>(new PushFilterDownTraverseRule());
+
+PushFilterDownTraverseRule::PushFilterDownTraverseRule() {
+  RuleSet::QueryRules().addRule(this);
+}
+
+const Pattern& PushFilterDownTraverseRule::pattern() const {
+  static Pattern pattern =
+      Pattern::create(PlanNode::Kind::kFilter,
+                      {Pattern::create(PlanNode::Kind::kAppendVertices,
+                                       {Pattern::create(PlanNode::Kind::kTraverse)})});
+  return pattern;
```

Comment:
Maybe we can implement this by splitting it into two rules: PushFilterDownAppendVertices and PushFilterIntoTraverse.

Reply:
Just mark it here. You can do refactoring later.

Reply:
Leave it for the future

Reply:
The following scenarios are not currently covered by this optimization:
```
match (v:player)-[e:like]->() where e.likeness>80  return e.likeness
```